prov/tcp/src/tcpx_ep.c

```diff
@@ -0,0 +1,769 @@
+
+	ret = getaddrinfo(sa_ip, sa_port, &hints, &addr_result);
+	if (ret) {
+		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+			"no available AF_INET address service:%s, %s\n",
+			sa_port, gai_strerror(ret));
+		return -FI_EINVAL;
+	}
+
+	for (iter = addr_result; iter; iter = iter->ai_next) {
+		_pep->cm.sock = socket(iter->ai_family, iter->ai_socktype,
+				      iter->ai_protocol);
+		if (_pep->cm.sock >= 0) {
+			tcpx_set_sockopts(_pep->cm.sock);
+			if (!bind(_pep->cm.sock, addr_result->ai_addr, addr_result->ai_addrlen))
+				break;
+			FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+				"failed to bind listener: f%s\n", strerror(errno));
+			ofi_close_socket(_pep->cm.sock);
+			_pep->cm.sock = -1;
+		}
+	}
+
+	freeaddrinfo(addr_result);
+
+	if (_pep->cm.sock < 0) {
+		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+			"failed to create listener: %s\n", strerror(errno));
+		return -FI_EIO;
+	}
+
+	if (listen(_pep->cm.sock, TCPX_MAX_SOCK_REQS)) {
+		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+			"socket listen failed\n");
+		goto out;
+	}
+
+	_pep->cm.do_listen = 1;
+
+	if (pthread_create(&_pep->cm.listener_thread, NULL,
+			   tcpx_cm_listener_thread, (void *)_pep)) {
+		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+			"failed to create cm thread\n");
+		goto out;
+	}
+
+	return FI_SUCCESS;
+ out:
+	ofi_close_socket(_pep->cm.sock);
+	_pep->cm.sock = -1;
+	return -FI_ENOMEM;
+}
+
+static struct fi_ops_cm tcpx_pep_cm_ops = {
+	.size = sizeof(struct fi_ops_cm),
+	.setname = fi_no_setname,
+	.getname = fi_no_getname,
+	.getpeer = fi_no_getpeer,
+	.connect = fi_no_connect,
+	.listen = tcpx_pep_listen,
+	.accept = fi_no_accept,
+	.reject = fi_no_reject,
+	.shutdown = fi_no_shutdown,
+	.join = fi_no_join,
+
+};
+
+
+static int tcpx_verify_info(uint32_t version, struct fi_info *info)
+{
+	return 0; //-FI_ENOSYS;
+
+}
+
+static struct fi_ops_ep tcpx_pep_ops = {
+	.size = sizeof(struct fi_ops_ep),
+	.getopt = fi_no_getopt,
+	.setopt = fi_no_setopt,
+	.tx_ctx = fi_no_tx_ctx,
+	.rx_ctx = fi_no_rx_ctx,
+	.rx_size_left = fi_no_rx_size_left,
+	.tx_size_left = fi_no_tx_size_left,
+};
+
+
+int tcpx_passive_ep(struct fid_fabric *fabric, struct fi_info *info,
+		    struct fid_pep **pep, void *context)
+{
+	int ret;
+	struct tcpx_pep *_pep;
+	struct addrinfo hints, *result;
+
+	if (info) {
+		ret = tcpx_verify_info(fabric->api_version, info);
+		if (ret) {
+			return ret;
+		}
+	}
+	_pep = calloc(1, sizeof(*_pep));
+	if (!_pep)
+		return -FI_ENOMEM;
+
+	if (info) {
+		_pep->info = *info;
+
+		if (info->src_addr) {
+			memcpy(&_pep->src_addr, info->src_addr,
+			       info->src_addrlen);
+		} else {
+			memset(&hints, 0, sizeof(hints));
+			hints.ai_family = AF_INET;
+			hints.ai_socktype = SOCK_STREAM;
+
+			ret = getaddrinfo("localhost", NULL, &hints, &result);
+			if (ret) {
+				ret = -FI_EINVAL;
+				FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,"getaddrinfo failed");
+				goto err;
+			}
+			memcpy(&_pep->src_addr, result->ai_addr,
+			       result->ai_addrlen);
+			freeaddrinfo(result);
+		}
```

Comment:
We should be able to allocate the socket here and bind it to the src_addr in fi_info.